module/Finna/src/Finna/View/CustomElement/AbstractBase.php

```diff
@@ -0,0 +1,199 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  CustomElements
+ * @author   Aleksi Peebles <aleksi.peebles@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:recommendation_modules Wiki
+ */
+namespace Finna\View\CustomElement;
+
+use Laminas\View\Model\ModelInterface;
+use Laminas\View\Model\ViewModel;
+use PHPHtmlParser\Dom;
+
+/**
+ * Abstract base custom element
+ *
+ * @category VuFind
+ * @package  CustomElements
+ * @author   Aleksi Peebles <aleksi.peebles@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:recommendation_modules Wiki
+ */
+abstract class AbstractBase implements CustomElementInterface
+{
+    /**
+     * Options
+     *
+     * The base class supports the following options:
+     *
+     * - attributes
+     *     Array of attributes for the custom element. These will overwrite
+     *     attributes parsed from outerHTML.
+     *
+     * - outerHTML
+     *     Outer HTML of the element. This will be parsed to a DOM object and
+     *     attributes.
+     *
+     * @var array
+     */
+    protected $options;
+
+    /**
+     * Attributes
+     *
+     * @var array
+     */
+    protected $attributes;
+
+    /**
+     * DOM object for the custom element if outerHTML is provided in options
+     *
+     * @var Dom
+     */
+    protected $dom = null;
+
+    /**
+     * View model for server-side rendering
+     *
+     * @var ModelInterface
+     */
+    protected $viewModel;
+
+    /**
+     * AbstractBase constructor.
+     *
+     * @param ?array $options       Options
+     * @param bool   $convertToBool Convert string true/false values to booleans.
+     */
+    public function __construct(?array $options = [], bool $convertToBool = false)
+    {
+        $options = $options ?? [];
+        $attributes = $options['attributes'] ?? [];
+
+        // If outer HTML is set, set up the DOM object and process attributes.
+        if (isset($options['outerHTML'])) {
+            $dom = (new Dom())->loadStr($options['outerHTML']);
+            if ($dom->countChildren() === 1
+                && $dom->firstChild()->getTag()->name() === $this->getName()
+            ) {
+                $this->dom = $dom;
+
+                // Attributes set in options overwrite attributes set in HTML.
+                $attributes = array_merge(
+                    $dom->firstChild()->getAttributes(), $attributes
+                );
+            }
```

Comment:
Pitäisikö käsitellä jotenkin tilanne, jossa outerHTML on annettu, mutta sen tagi ei vastaa `$this->getName()`:a?

Reply:
Sen on tarkoitus olla huomioituna tuossa `if ($dom->countChildren() === 1 && $dom->firstChild()->getTag()->name() === $this->getName()`, eli outerHTML käsitellään vain jos vastaa nimeä ja muutenkin vaikuttaa järkevältä. Toinen juttu sitten että pitäisikö siitä seurata ihan virhe?

Reply:
Sopii hyvin. Silloin mahdolliset ongelmat eivät jää piiloon.